src/algorithms/cache.rs

```diff
@@ -0,0 +1,299 @@
+            let reference = Correctness::pack(&last.mask);
+            let last_idx = self
+                .remaining
+                .iter()
+                .find(|(word, _, _)| &*last.word == *word)
+                .unwrap()
+                .2;
+            COMPUTES.with(|c| {
+                let row = &c.get().unwrap()[last_idx];
+                if matches!(self.remaining, Cow::Owned(_)) {
+                    self.remaining.to_mut().retain(|(word, _, word_idx)| {
+                        reference == get_correctness_packed(row, &last.word, word, *word_idx)
+                    });
+                } else {
+                    self.remaining = Cow::Owned(
+                        self.remaining
+                            .iter()
+                            .filter(|(word, _, word_idx)| {
+                                reference
+                                    == get_correctness_packed(row, &last.word, word, *word_idx)
+                            })
+                            .copied()
+                            .collect(),
+                    );
+                }
+            });
+        }
+        if history.is_empty() {
+            self.patterns = Cow::Borrowed(PATTERNS.get().unwrap());
+            // NOTE: I did a manual run with this commented out and it indeed produced "tares" as
+            // the first guess. It slows down the run by a lot though.
+            return "tares".to_string();
+        } else {
+            assert!(!self.patterns.is_empty());
+        }
+
+        let remaining_p: f64 = self.remaining.iter().map(|&(_, p, _)| p).sum();
+        let remaining_entropy = -self
+            .remaining
+            .iter()
+            .map(|&(_, p, _)| {
+                let p = p / remaining_p;
+                p * p.log2()
+            })
+            .sum::<f64>();
+        self.entropy.push(remaining_entropy);
+
+        let mut best: Option<Candidate> = None;
+        let mut i = 0;
+        let stop = (self.remaining.len() / 3).max(20);
+        for &(word, count, word_idx) in &*self.remaining {
+            // considering a world where we _did_ guess `word` and got `pattern` as the
+            // correctness. now, compute what _then_ is left.
+
+            // Rather than iterate over the patterns sequentially and add up the counts of words
+            // that result in that pattern, we can instead keep a running total for each pattern
+            // simultaneously by storing them in an array. We can do this since each candidate-word
+            // pair deterministically produces only one mask.
+            let mut totals = [0.0f64; MAX_MASK_ENUM];
+
+            COMPUTES.with(|c| {
+                let row = &c.get().unwrap()[word_idx];
+                for (candidate, count, candidate_idx) in &*self.remaining {
+                    let idx = get_correctness_packed(row, word, candidate, *candidate_idx);
+                    totals[idx as usize] += count;
```

Comment:
I _think_ clippy will complain about using a cast here and will recommend `usize::from` instead so that future changes to the type of `idx` won't accidentally result in this becoming a lossy cast.